### ISSUE TITLE
A derived class is allowed be created without a parent's getter or setter.

### ISSUE THREAD

**Issue by dwilsonactual** [microsoft/TypeScript#47581]
# Bug Report

### 🔎 Search Terms

class extend override getter setter accessor

### 🕗 Version & Regression Information

- This seems to occur on all versions on the Playground (3.9+)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/MYGwhgzhAEAKYCcCmA7ALtA3gKGn6ADgK4BGIAlsNAOZIYBmAFAJRbTJpEIrQvQC8APiwBfANzQR2KdlCQYAYQAW5EABNoSAB5pUamPGTosufMTKVoEOtCb0iKYAC5erIdABuAe3JrWmSWlsWS8UCAwqfmgUJAB3aGVVNRYxWQA6JmYJAHps9gc0cgBbJE0EBC8EaBIkYDAiawSVdWg1LyQIFAByDGKCECQS4zQlUto0XSr6SttsIA)

### 💻 Code

```ts
class Parent {
    public get f() { return () => {}; }
}

class Child extends Parent {
    public set f(func: () => void) { }
}

const c = new Child();
c.f(); // runtime error because Child doesn't implement the getter for f
```

### 🙁 Actual behavior

Class `Child` is allowed to extend class `Parent` even though, by adding a setter for `f`, it discards the getter for `f` and therefore doesn't implement Parent's contract.

### 🙂 Expected behavior

TypeScript should emit an error because `Child` is not compatible with `Parent`.

**Comment by MartinJohns** [comment]
Sounds like #30852.

**Comment by dwilsonactual** [comment]
@MartinJohns I think you're right, the root of the problem is that when you add a setter, TypeScript assumes you have a getter whether one exists or not, and this is apparently by design.

**Comment by RyanCavanaugh** [comment]
It'd be nice to fix this but I'm not even sure how. The problem is that this isn't a soundness violation -- if we write out the getter that is implicitly present, there's no problem there either:
```ts
class Parent {
    public get f() { return () => {}; }
}

class Child extends Parent {
    public get(): never {
        throw new Error("readonly");
    }
    public set f(func: () => void) { }
}
```
An ad-hoc rule that a setter must have a matching getter if there's a property in the base class feels like a hack.

**Comment by OwenDelahoy** [comment]
In the following code, typescript expects a boolean, however at runtime it is actually undefined.

[Playground Link](https://www.typescriptlang.org/play?ts=4.5.4#code/MYGwhgzhAEAKYCcCmA7ALtA3gKGn6ADggPZpLBkAm0A+gG5ggCuS0AvNAGaMRIDcufASYAjEAEtg0AOZIMDZkgAUASizRkaJghTQ0AC3EQAdPUYs+0AL7Yb2UJBgBhQyGpIAHmRSUY8ZOhYgnjCYpLQvPLmygosAFzQIsTEIEhgKGqYeoYmZors0LH81rbY2JxMKBTixLqcqglJKWm6OPjQwLUQGMQiAFYFKEgA7tAu4m6qAu2a2rq9fcZFAnYVVWg1utINeggsQe2dKN3QC4MjY66UU8Gn-UvRBWh7-LezOneLRaX2Xc3GIGI23qKjUAHowdBKpQkJxxENKL9jv9AdttqDoBCoT5YfCkIigA)
```TypeScript
class Parent {
    protected _value = false;
    public get value() { return this._value; }
}

class Child extends Parent {
    public set value(value: boolean) { this._value = value; }
}

function f(): boolean {
    const obj = new Child();
    return obj.value;
}

function g(): true {
    const obj = new Child();
    obj.value = true;
    return obj.value
}

console.log(f()) // undefined
console.log(g()) // undefined

```

**Comment by BlakeOne** [comment]
> I think that most would expect `Child` to inherit `Parent`'s getter

Agree completely that it’s expected to inherit the getter. Inheriting everything that isn’t overridden is expected. IMO this is a bug which makes overriding the setter unnecessarily difficult.

**Comment by Jack-Works** [comment]
just being hit by this issue. The typeScript should error if the parent class has getter/setter but the child only have setter, because in most cases, people's intention is to "inherit" the getter from the parent class.

```ts
class Parent {
    get a() {
        return 1
    }
    set a(val: number) {}
}
class Child extends Parent {
    set a(val: number) {
        super.a = val
    }
}
const child = new Child()
child.a = 2
console.log('--- after write --')
console.log(child.a)
//          ts: child.a is number
//          js: child.a is undefined

```

**Comment by viktorpts** [comment]
I'm joining this discussion, since if this behavior is by design, it's the same as reducing the accessibility of an overriden property compared to the base class (you're essentially creating a private getter). This not only breaks Liskov Substitution, but is also a compilation error if you attempt to do it explicitly:

```javascript
class Parent {
    get a() { ... }
    set a(val: number) { ... }
}

class Child extends Parent {
    private override get a() { ... } // Compilation error
}
```

The child class should either have access to the parent's getter, or this code should throw a compilation error.

**Comment by 54145a** [comment]
This is a **Bug**, not some "Experience Enhancement". The issue is not about developer convenience or style; it is about a fundamental breakdown in type safety.

When a subclass overrides only the setter of a parent getter/setter pair, the parent’s getter is silently discarded at runtime. This breaks the **substitutability** principle: a `Child` instance is no longer a valid `Parent`.

```ts
class Parent {
    get value(): string { return "parent"; }
    set value(v: string) { console.log("Setter called."); }
}
class Child extends Parent {
    set value(v: string) {}
}
const p: Parent = new Child();
console.log(p.value); // undefined
```

From the perspective of `Parent`’s contract, `Child` violates it: the `value` property is expected to be a readable `string`, but it becomes `undefined`.

While it’s true that JavaScript’s property descriptor mechanism causes this behavior, TypeScript’s role is to model and constrain JavaScript semantics to catch common errors. Allowing a subclass to silently break its parent’s contract is inconsistent with TypeScript’s usual strictness.

TypeScript should at least emit an error when a subclass overrides only the setter but the parent has a getter&setter pair. This would not change JavaScript runtime behavior; it would only add a compile-time check that prevents a common source of bugs.
